src/ipcalc.1
.TH IPCALC 1 "April 30 2001" "Red Hat, Inc." \" -*- nroff -*-
.SH NAME
ipcalc \- perform simple manipulation of IP addresses
.SH SYNOPSIS
.B ipcalc
[\fIOPTION\fR]... <\fBIP address\fR>[\fI/prefix\fR] [\fInetmask\fR]

.SH DESCRIPTION
\fBipcalc\fR provides a simple way to calculate IP information for a host.
The various options specify what information \fBipcalc\fR should display
on standard out. Multiple options may be specified.  An IP address to
operate on must always be specified.  Most operations also require a
netmask or a CIDR prefix as well.

.SH OPTIONS
.TP
.TP
\fB\-c\fR, \fB\-\-check\fR
Validate the IP address under the specified family.  If no address
family is specified, IPv4 is assumed.

.TP
\fB\-4\fR, \fB\-\-ipv4\fR
Specify IPv4 address family (default).

.TP
\fB\-6\fR, \fB\-\-ipv6\fR
Specify IPv6 address family.

.TP
\fB\-b\fR, \fB\-\-broadcast\fR
Display the broadcast address for the given IP address and netmask.

.TP
\fB\-h\fR, \fB\-\-hostname\fR
Display the hostname for the given IP address.

.TP
\fB\-m\fR, \fB\-\-netmask\fR
Calculate the netmask for the given IP address. It assumes that the IP
address is in a complete class A, B, or C network. Many networks do
not use the default netmasks, in which case an inappropriate value will
be returned.

.TP
\fB\-p\fR, \fB\-\-prefix\fR
Show the prefix for the given mask/IP address.

.TP
\fB\-n\fR, \fB\-\-network\fR 
Display the network address for the given IP address and netmask.

.TP
\fB\-s\fR, \fB\-\-silent\fR 
Don't ever display error messages.

.SH AUTHORS
.nf
Erik Troan <ewt@redhat.com>
.nf
Preston Brown <pbrown@redhat.com>

.TP
IPv6 supported wedged in by David Cantrell <dcantrell@redhat.com>
.fi
.SH "REPORTING BUGS"
Report bugs at http://bugzilla.redhat.com/
.SH COPYRIGHT
Copyright \(co 1997-2008 Red Hat, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.


ChangeLog
commit 6a6e9b2d8f9dfd1f8acee86b8c835d008d5e6007
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 11:42:48 2008 +0200

    ignore %{?dist} in release

 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 427db7a82925a46c080c100dc9954fe1ecb8c949
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 09:34:02 2008 +0200

    add %{?dist} to release
    changed URL
    refined License tag
    add changelog

 initscripts.spec |   17 +++++++++++++----
 1 files changed, 13 insertions(+), 4 deletions(-)

commit c9329c4aa3a4e00c2e56a1556e8438e80e99f425
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Mar 10 15:23:37 2008 +0100

    only create ChangLog from current branch

 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 941a2a83127f206aa46c0e55cc57eada091c7abe
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 10:23:22 2008 +0200

    https://bugzilla.redhat.com/show_bug.cgi?id=433702
    
    Changes to init scripts to better support stateless:
    
    --rc.sysinit--
    
     - don't update mtab in a r/o root
     - mount on sys and /dev/pts returns EBUSY on a readonly root
     - [ -n $CLIENTSTATE ] is always returning true, causing an unnecessary timeout
    
    --network--
    
     - /etc/mtab may not have any mount information in the readonly-root
       cause, but the root mount is likely to be a network mount.  Without
       this, nfs hangs the reboot process.
    
    --halt--
    
     - don't try and update a read-only mtab
    
    --rwtab--
    
     - /var/empty/sshd/etc/localtime needs to be overwritten by sshd at
       startup
     - /var/lib/random-seed is written to at shutdown; this fixes the
       error but doesn't return the functionality lost in a r/o /var/lib.

 rc.d/init.d/halt    |    9 ++++++---
 rc.d/init.d/network |    2 +-
 rc.d/rc.sysinit     |   30 ++++++++++++++++++------------
 rwtab               |    2 ++
 4 files changed, 27 insertions(+), 16 deletions(-)

commit 58ddef0ad76b02840d6cf8babb0d036f1a1f1c83
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 10:04:35 2008 +0200

    https://bugzilla.redhat.com/show_bug.cgi?id=452866
    
    The _netdev option is used in /etc/fstab to mark devices that depend on the
    network for operation. This causes the initscripts to defer mounting file
    systems located on these devices until the network is brought up and the netfs
    initscript runs.
    
    Device nodes for these devices (nbd, gnbd, iscsi etc.) are created by udev once
    the appropriate subsystem is started.
    
    Processing of the udev backlog may not have completed by the time the netfs
    script runs causing failures to occur when attempting to mount file systems.
    This can happen when there are a large number of devices or when device
    initialisation is slow.
    
    This is fixed by adding a udevsettle to the start() function in netfs.

 rc.d/init.d/netfs |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit a31538409a960560f35494ff30a6181f16d1d6de
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 10:02:51 2008 +0200

    https://bugzilla.redhat.com/show_bug.cgi?id=447928
    
    All networking scripts from /etc/sysconfig/network-scripts call logger without
    full path. This can prevent system from booting if let's say dcc is installed
    when compiled with wrong parameters and  installs /usr/sbin/logger (see bellow)
    It's not really a bug but this can be easily prevented. All /etc/init.d scripts
    use full path for logger network-scripts should use it also

 rc.d/init.d/netconsole                           |    2 +-
 sysconfig/network-scripts/ifdown-sl              |    2 +-
 sysconfig/network-scripts/ifup                   |    6 +++---
 sysconfig/network-scripts/ifup-ippp              |   14 +++++++-------
 sysconfig/network-scripts/ifup-ppp               |   10 +++++-----
 sysconfig/network-scripts/ifup-sl                |    6 +++---
 sysconfig/network-scripts/network-functions      |    2 +-
 sysconfig/network-scripts/network-functions-ipv6 |    6 +++---
 8 files changed, 24 insertions(+), 24 deletions(-)

commit cefaaa0dd6b2d60f674875cc8b86d5846e8033f7
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Sep 16 09:41:16 2008 +0200

    https://bugzilla.redhat.com/show_bug.cgi?id=437110
    
    On System z, typical hardware setups can consist of up to several thousand I/O
    devices. The current udev rules for handling these devices and the ccw_init
    helper script are implemented in an insufficient way, leading to unnecessarily
    high usage of memory and cpu time during system startup.
    
    The attached patch against the initscripts-8.45.19.EL-1 source RPM reduces
    this resource usage drastically by
    
    a) fixing a number of coding problems in helper script ccw_init (the current
    version creates failure indications in /dev/.udev/failed for each device)
    b) restricting the ccw_init script to network devices (ccw_init is only
    relevant for network devices)
    
    Using these changes, memory consumption on an example system with ~1500
    devices could be reduced by 80% (VSZ), resp. 50% (RSS) and the elapsed
    duration for udev could be reduced from 19s down to 6 seconds.
    
    Business case: with reduced memory and time consumption during startup, more
    Linux images can be run with the same amount of resources.

 src/ccw_init              |   13 +++++++------
 udev/rules.d/55-ccw.rules |   10 +++++++++-
 2 files changed, 16 insertions(+), 7 deletions(-)

commit c0e42318dcca68d960ca1d7977af2979bc99b8aa
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Apr 1 13:04:12 2008 -0400

    don't attempt to re-enslave already-enslaved devices (#440077)

 sysconfig/network-scripts/ifup-eth |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

commit 8b1672cd138cca0e60bccdc6317f5ad4dff16d4c
Author: Bill Nottingham <notting@redhat.com>

src/ipcalc.c
/*
 * Copyright (c) 1997-2008 Red Hat, Inc. All rights reserved.
 *
 * This software may be freely redistributed under the terms of the GNU
 * public license.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Authors:
 *   Erik Troan <ewt@redhat.com>
 *   Preston Brown <pbrown@redhat.com>
 *   David Cantrell <dcantrell@redhat.com>
 */

#include <ctype.h>
#include <popt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

/*!
  \file ipcalc.c
  \brief provides utilities for manipulating IP addresses.

  ipcalc provides utilities and a front-end command line interface for
  manipulating IP addresses, and calculating various aspects of an ip
  address/netmask/network address/prefix/etc.

  Functionality can be accessed from other languages from the library
  interface, documented here.  To use ipcalc from the shell, read the
  ipcalc(1) manual page.

  When passing parameters to the various functions, take note of whether they
  take host byte order or network byte order.  Most take host byte order, and
  return host byte order, but there are some exceptions.
*/

/*!
  \fn struct in_addr prefix2mask(int bits)
  \brief creates a netmask from a specified number of bits

  This function converts a prefix length to a netmask.  As CIDR (classless
  internet domain internet domain routing) has taken off, more an more IP
  addresses are being specified in the format address/prefix
  (i.e. 192.168.2.3/24, with a corresponding netmask 255.255.255.0).  If you
  need to see what netmask corresponds to the prefix part of the address, this
  is the function.  See also \ref mask2prefix.

  \param prefix is the number of bits to create a mask for.
  \return a network mask, in network byte order.
*/
struct in_addr prefix2mask(int prefix) {
    struct in_addr mask;
    memset(&mask, 0, sizeof(mask));
    mask.s_addr = htonl(~((1 << (32 - prefix)) - 1));
    return mask;
}

/*!
  \fn int mask2prefix(struct in_addr mask)
  \brief calculates the number of bits masked off by a netmask.

  This function calculates the significant bits in an IP address as specified by
  a netmask.  See also \ref prefix2mask.

  \param mask is the netmask, specified as an struct in_addr in network byte order.
  \return the number of significant bits.  */
int mask2prefix(struct in_addr mask)
{
    int i;
    int ipbits = sizeof(struct in_addr) * 8;
    int count = ipbits;
    uint32_t saddr = mask.s_addr;

    for (i = 0; i < ipbits; i++) {
        if (!(ntohl(saddr) & ((2 << i) - 1)))
            count--;
    }

    return count;
}

/*!
  \fn struct in_addr default_netmask(struct in_addr addr)

  \brief returns the default (canonical) netmask associated with specified IP
  address.

  When the Internet was originally set up, various ranges of IP addresses were
  segmented into three network classes: A, B, and C.  This function will return
  a netmask that is associated with the IP address specified defining where it
  falls in the predefined classes.

  \param addr an IP address in network byte order.
  \return a netmask in network byte order.  */
struct in_addr default_netmask(struct in_addr addr)
{
    uint32_t saddr = addr.s_addr;
    struct in_addr mask;

    memset(&mask, 0, sizeof(mask));

    if (((ntohl(saddr) & 0xFF000000) >> 24) <= 127)
        mask.s_addr = htonl(0xFF000000);
    else if (((ntohl(saddr) & 0xFF000000) >> 24) <= 191)
        mask.s_addr = htonl(0xFFFF0000);
    else
        mask.s_addr = htonl(0xFFFFFF00);

    return mask;
}

/*!
  \fn struct in_addr calc_broadcast(struct in_addr addr, int prefix)

  \brief calculate broadcast address given an IP address and a prefix length.

  \param addr an IP address in network byte order.
  \param prefix a prefix length.

  \return the calculated broadcast address for the network, in network byte
  order.
*/
struct in_addr calc_broadcast(struct in_addr addr, int prefix)
{
    struct in_addr mask = prefix2mask(prefix);
    struct in_addr broadcast;

    memset(&broadcast, 0, sizeof(broadcast));
    broadcast.s_addr = (addr.s_addr & mask.s_addr) | ~mask.s_addr;
    return broadcast;
}

/*!
  \fn struct in_addr calc_network(struct in_addr addr, int prefix)
  \brief calculates the network address for a specified address and prefix.

  \param addr an IP address, in network byte order
  \param prefix the network prefix
  \return the base address of the network that addr is associated with, in
  network byte order.
*/
struct in_addr calc_network(struct in_addr addr, int prefix)
{
    struct in_addr mask = prefix2mask(prefix);
    struct in_addr network;

    memset(&network, 0, sizeof(network));
    network.s_addr = addr.s_addr & mask.s_addr;
    return network;
}

/*!
  \fn const char *get_hostname(int family, void *addr)
  \brief returns the hostname associated with the specified IP address

  \param family the address family, either AF_INET or AF_INET6.
  \param addr an IP address to find a hostname for, in network byte order,
  should either be a pointer to a struct in_addr or a struct in6_addr.

  \return a hostname, or NULL if one cannot be determined.  Hostname is stored
  in a static buffer that may disappear at any time, the caller should copy the
  data if it needs permanent storage.
*/
char *get_hostname(int family, void *addr)
{
    struct hostent * hostinfo = NULL;
    int x;
    struct in_addr addr4;
    struct in6_addr addr6;

    if (family == AF_INET) {
        memset(&addr4, 0, sizeof(addr4));
        memcpy(&addr4, addr, sizeof(addr4));
        hostinfo = gethostbyaddr((const void *) &addr4,
                                 sizeof(addr4), family);
    } else if (family == AF_INET6) {
        memset(&addr6, 0, sizeof(addr6));
        memcpy(&addr6, addr, sizeof(addr6));
        hostinfo = gethostbyaddr((const void *) &addr6,
                                 sizeof(addr6), family);
    }

    if (!hostinfo)
        return NULL;

    for (x=0; hostinfo->h_name[x]; x++) {
        hostinfo->h_name[x] = tolower(hostinfo->h_name[x]);
    }
    return hostinfo->h_name;
}

/*!
  \fn main(int argc, const char **argv)
  \brief wrapper program for ipcalc functions.

  This is a wrapper program for the functions that the ipcalc library provides.
  It can be used from shell scripts or directly from the command line.

  For more information, please see the ipcalc(1) man page.
*/
int main(int argc, const char **argv) {
    int showBroadcast = 0, showPrefix = 0, showNetwork = 0;
    int showHostname = 0, showNetmask = 0;
    int beSilent = 0;
    int doCheck = 0, familyIPv4 = 0, familyIPv6 = 0;
    int rc;
    poptContext optCon;
    char *ipStr, *prefixStr, *netmaskStr, *chptr;
    char *hostName = NULL;
    char namebuf[INET6_ADDRSTRLEN+1];
    struct in_addr ip, netmask, network, broadcast;
    struct in6_addr ip6;
    int prefix = 0;
    char errBuf[250];
    struct poptOption optionsTable[] = {
        { "check", 'c', 0, &doCheck, 0,
          "Validate IP address for specified address family", },
        { "ipv4", '4', 0, &familyIPv4, 0,
          "IPv4 address family (default)", },
        { "ipv6", '6', 0, &familyIPv6, 0,
          "IPv6 address family", },
        { "broadcast", 'b', 0, &showBroadcast, 0,
          "Display calculated broadcast address", },
        { "hostname", 'h', 0, &showHostname, 0,
          "Show hostname determined via DNS" },
        { "netmask", 'm', 0, &showNetmask, 0,
          "Display default netmask for IP (class A, B, or C)" },
        { "network", 'n', 0, &showNetwork, 0,
          "Display network address", },
        { "prefix", 'p', 0, &showPrefix, 0,
          "Display network prefix", },
        { "silent", 's', 0, &beSilent, 0,
          "Don't ever display error messages" },
        POPT_AUTOHELP
        { NULL, '\0', 0, 0, 0, NULL, NULL }
    };

    optCon = poptGetContext("ipcalc", argc, argv, optionsTable, 0);
    poptReadDefaultConfig(optCon, 1);

    if ((rc = poptGetNextOpt(optCon)) < -1) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: bad argument %s: %s\n",
                    poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
                    poptStrerror(rc));
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    if (!(ipStr = (char *) poptGetArg(optCon))) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: ip address expected\n");
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    if (strchr(ipStr,'/') != NULL) {
        prefixStr = strchr(ipStr, '/') + 1;
        prefixStr--;
        *prefixStr = '\0';  /* fix up ipStr */
        prefixStr++;
    } else
        prefixStr = NULL;

    if (prefixStr != NULL) {
        prefix = atoi(prefixStr);
        if (prefix == 0) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad prefix: %s\n", prefixStr);
            return 1;
        }
    }

    if (showBroadcast || showNetwork || showPrefix) {
        if (!(netmaskStr = (char *) poptGetArg(optCon)) && (prefix == 0)) {
            if (!beSilent) {
                fprintf(stderr, "ipcalc: netmask or prefix expected\n");
                poptPrintHelp(optCon, stderr, 0);
            }
            return 1;
        } else if (netmaskStr && prefix != 0) {
            if (!beSilent) {
                fprintf(stderr, "ipcalc: both netmask and prefix specified\n");
                poptPrintHelp(optCon, stderr, 0);
            }
            return 1;
        } else if (netmaskStr) {
            if (inet_pton(AF_INET, netmaskStr, &netmask) <= 0) {
                if (!beSilent)
                    fprintf(stderr, "ipcalc: bad netmask: %s\n", netmaskStr);
                return 1;
            }
            prefix = mask2prefix(netmask);
        }
    }

    if ((chptr = (char *) poptGetArg(optCon))) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: unexpected argument: %s\n", chptr);
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    /* Handle CIDR entries such as 172/8 */
    if (prefix) {
        char *tmp = ipStr;
        int i;

        for (i=3; i> 0; i--) {
            tmp = strchr(tmp,'.');
            if (!tmp)
                break;
            else
                tmp++;
        }

        tmp = NULL;
        for (; i>0; i--) {
            if (asprintf(&tmp, "%s.0", ipStr) == -1) {
                fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
                abort();
            }
            ipStr = tmp;
        }
    }

    if (!familyIPv4 && !familyIPv6)
        familyIPv4 = 1;

    if (familyIPv4 && familyIPv6) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: cannot specify both address families\n");
        }
        return 1;
    }

    if (familyIPv4) {
        if (inet_pton(AF_INET, ipStr, &ip) <= 0) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv4 address: %s\n", ipStr);
            return 1;
        } else {
            if (doCheck)
                return 0;
        }
    }

    if (familyIPv6) {
        if (inet_pton(AF_INET6, ipStr, &ip6) <= 0) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", ipStr);
            return 1;
        } else {
            if (doCheck)
                return 0;
        }
    }

    if (familyIPv6 &&
        (showBroadcast || showNetmask || showNetwork || showPrefix)) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: unable to show setting for IPv6\n");
        }
        return 1;
    }

    if (familyIPv4 &&
        !(showNetmask|showPrefix|showBroadcast|showNetwork|showHostname)) {
        poptPrintHelp(optCon, stderr, 0);
        return 1;
    }

    poptFreeContext(optCon);

    /* we know what we want to display now, so display it. */

    if (showNetmask) {
        if (prefix) {
            netmask = prefix2mask(prefix);
        } else {
            netmask = default_netmask(ip);
            prefix = mask2prefix(netmask);
        }

        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &netmask, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort();
        }

        printf("NETMASK=%s\n", namebuf);
    }

    if (showPrefix) {
        if (!prefix)
            prefix = mask2prefix(ip);
        printf("PREFIX=%d\n", prefix);
    }

    if (showBroadcast) {
        broadcast = calc_broadcast(ip, prefix);
        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &broadcast, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort();
        }

        printf("BROADCAST=%s\n", namebuf);
    }

    if (showNetwork) {
        network = calc_network(ip, prefix);
        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &network, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort(); 
        }

        printf("NETWORK=%s\n", namebuf);
    }

    if (showHostname) {
        if (familyIPv4) {
            hostName = get_hostname(AF_INET, &ip);
        } else if (familyIPv6) {
            hostName = get_hostname(AF_INET6, &ip6);
        }

        if (hostName == NULL) {
            if (!beSilent) {
                sprintf(errBuf, "ipcalc: cannot find hostname for %s", ipStr);
                herror(errBuf);
            }
            return 1;
        }

        printf("HOSTNAME=%s\n", hostName);
    }

    return 0;
}
